### PR TITLE
Optimizing MAP metric

### PR DESCRIPTION
**Pull request by romanie**

### CODE REVIEW

File: diplom.py [hunk 1]
```diff
@@ -21,14 +21,17 @@ def mean_average_precision(predicted_probability_distribution,
     k = -1
     for row in predicted_probability_distribution:
```
Comment (romanie) [inline diplom.py:22]: Use [enumerate](https://docs.python.org/3/library/functions.html#enumerate) here for k please.

File: diplom.py [hunk 2]
```diff
@@ -21,14 +21,17 @@ def mean_average_precision(predicted_probability_distribution,
...
+        place = 1
         for i in xrange(0, len(list_of_classes)):
-            classes_by_prob.append([row[i], list_of_classes[i]])
-        classes_by_prob.sort(reverse=True)    
-        for j in range (0, min(5, len(classes_by_prob))):
-            if classes_by_prob[j][1] == test_y[k]:
-                total_error += 1.0 / (j + 1)
-    
+            if list_of_classes[i] == test_y[k]:
+                prob_of_class = row[i]
```
Comment (romanie) [inline diplom.py]: rename `prob_of_class` to `prob_of_target_class`

File: diplom.py [hunk 3]
```diff
@@ -21,14 +21,17 @@ def mean_average_precision(predicted_probability_distribution,
...
+                break
+        for i in xrange(0, len(list_of_classes)):
+            if row[i] > prob_of_class or (row[i] == prob_of_class and list_of_classes[i] > test_y[k]):
```
Comment (romanie) [inline diplom.py]: Add a newline before 'or'. 

Note that you'll need to add parenthesis to formulate a multiline expression - `if a or b` => `if (a or b)`.

File: diplom.py [hunk 4]
```diff
@@ -21,14 +21,17 @@ def mean_average_precision(predicted_probability_distribution,
...
+                place += 1
+        if place < 6:
+            total_error += 1.0 / (place)
```
Comment (romanie) [inline diplom.py]: You don't need parenthesis around `place`.

File: diplom.py (hunk 1 above)
Comment (oksanaie) [inline diplom.py:22]: I don't understand how to use it
list(enumeratirs(predicted_probability_distribution))
and then each time _next_() ?

File: diplom.py [hunk 5]
```diff
@@ -18,17 +18,19 @@ def mean_average_precision(predicted_probability_distribution,
                            list_of_classes, 
                            test_y):
     total_error = 0
-    k = -1
-    for row in predicted_probability_distribution:
-        k += 1
-        classes_by_prob = []
+    for k,row in (enumerate (predicted_probability_distribution)):
```
Comment (romanie) [inline diplom.py]: `k,row` -> `k, row`
also, you don't need parenthesis around enumerate

File: diplom.py [hunk 6]
```diff
@@ -18,17 +18,19 @@ def mean_average_precision(predicted_probability_distribution,
...
+        prob_of_target_class = 0
+        place = 1
         for i in xrange(0, len(list_of_classes)):
```
Comment (romanie) [inline diplom.py]: Since in next two loops you implicitly rely on the fact that `len(list_of_classes) == len(row)` I suggest you adding an [assertion](http://www.tutorialspoint.com/python/assertions_in_python.htm) before first for, i.e.:
`assert len(list_of_classes) == len(row)`

File: diplom.py [hunk 7]
```diff
@@ -18,28 +18,33 @@ def mean_average_precision(predicted_probability_distribution,
...
+                place += 1
+        if place < 6:
+            total_error += 1.0 / place
     return total_error / len(predicted_probability_distribution)
 
 start_time = time.time()
 
 parser = argparse.ArgumentParser()
 parser.add_argument("--model", choices=["random_forest", "k_neighbors", "logistic_regression"], default=["random_forest"], dest="model")
+parser.add_argument("--dataset_size", default=100000, type=int, dest="dataset_size")
```
Comment (romanie) [inline diplom.py]: Please add two flags: train_dataset_size and test_dataset_size. Whole point of this change is to allow disabling "test" part of the script so we could run training and save model to the file without running any benchmarking. 

Although useful for a quick iteration on the model, the end state that you want to achieve is to have two scripts:
- train_model.py [will be grounded on the current diploma.py, will allow you to train model on the data, cross-validate ("test") on a portion of it and save the trained model to a file]  
- apply_model.py [will be another script that will reuse data preprocessing from diploma.py and run prediction model saved by first script to prepare predictions for the Kaggle test file]

With this infrastructure you will be all set to submit a first version to Kaggle and continue your work on adding features and visualization of process.

File: diplom.py [hunk 8]
```diff
@@ -18,28 +18,33 @@ def mean_average_precision(predicted_probability_distribution,
...
 args = parser.parse_args()
 print "Model used in this run is %s" % args.model
+print "Dataset size is %s" % args.dataset_size
 
-TRAIN_DATASET_SIZE = 300000
-TEST_DATASET_SIZE = 300000
+TRAIN_DATASET_SIZE = args.dataset_size
```
Comment (romanie) [inline diplom.py]: Please don't use intermediate variables here.

File: diplom.py [hunk 9]
```diff
@@ -18,28 +18,33 @@ def mean_average_precision(predicted_probability_distribution,
...
+        place = 1
+        assert len(list_of_classes) == len(row)
+        for i in xrange(0, len(row)):
+            if list_of_classes[i] == test_y[k]:
+                prob_of_target_class = row[i]
+                break
+        for i in xrange(0, len(row)):
+            if (row[i] > prob_of_target_class 
+                or (row[i] == prob_of_target_class and list_of_classes[i] > test_y[k])):
+                place += 1
```
Comment (romanie) [inline diplom.py:34]: Could you also add here `if place >= 6: break`
